### ISSUE TITLE
PHP "float" serialization loses precision

### ISSUE THREAD

**Issue by bshaffer** [protocolbuffers/protobuf#8165]
In both the protobuf native PHP library and PHP C-extension, I am seeing a loss of precision with floats:

```
message TestMessage {
  float foo = 1;
}
```

```php
$initialMessage = new TestMessage();
$initialMessage->setFoo(-8.6107481E7);
echo "Initial message: ";
var_dump($initialMessage->getFoo());

echo "Serialized message: ";
$serializedMessage = new TestMessage();
$serializedMessage->mergeFromString($initialMessage->serializeToString());
var_dump($serializedMessage->getFoo());
```

When the C-Extension is enabled, the loss of precision happens immediately. When the C-extension is not enabled, the loss of precision happens when the message is serialized and deserialized:

```
// Without protobuf C-Extension
Initial message: float(-86107481)
Serialized message: float(-86107480)

// With protobuf C-Extension
Initial message: float(-86107480)
Serialized message: float(-86107480)
```

**Comment by acozzette** [comment]
I haven't done the math carefully but this sounds like it could be just caused by the fact that the protobuf `float` type is only 32 bits.  As I understand, PHP always uses double-precision floats and that would explain why the plain PHP implementation doesn't lose precision by itself but only after a serialize and parse. The C extension presumably distinguishes between `float` and `double` and therefore converts immediately to the 32-bit `float`, resulting in the immediate precision loss.

**Comment by haberman** [comment]
Yes, Adam is correct. The bug here is with pure-PHP, which should truncate this to float precision in the setter.

**Comment by bshaffer** [comment]
@haberman So are we saying this is an issue in the Pure-PHP protobuf library, and the way to fix it is to update float setters to truncate to single precision?

**Comment by bshaffer** [comment]
I added a fix in https://github.com/protocolbuffers/protobuf/pull/8187 that passes for my test suites

**Comment by haberman** [comment]
This was fixed in #8187.
